multiview/embed/gcca.py

```diff
@@ -0,0 +1,251 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .base import BaseEmbed
+from ..utils.utils import check_Xs
+
+import numpy as np
+from scipy import linalg, stats
+from scipy.sparse.linalg import svds
+from sklearn.preprocessing import normalize
+
+
+class GCCA(BaseEmbed):
+    """
+    An implementation of Generalized Canonical Correalation Analysis.Computes
+    individual projections into a common subspace such that the correlations
+    between pairwise projections are minimized (ie. maximize pairwise
+    correlation). Reduces to CCA in the case of two samples.
+
+    See https://www.sciencedirect.com/science/article/pii/S1053811912001644?
+    via%3Dihub
+    for relevant details.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._projection_mats = None
+
+    def _preprocess(self, X):
+        """
+        Subtracts the row means and divides by the row standard deviations.
+        Then subtracts column means.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_observations, n_features)
+            The data to preprocess
+
+        Returns
+        -------
+        X2 : preprocessed data matrix
+        """
+
+        # Mean along rows using sample mean and sample std
+        X2 = stats.zscore(X, axis=1, ddof=1)
+        # Mean along columns
+        mu = np.mean(X2, axis=0)
+        X2 -= mu
+        return X2
+
+    def _check_inputs(self, fraction_var, sv_tolerance, n_components, size):
+        if fraction_var is None:
+            pass
+        elif not isinstance(fraction_var, float) and not isinstance(
+            fraction_var, int
+        ):
+            raise TypeError("fraction_var must be an integer or float")
+        elif fraction_var <= 0 or fraction_var > 1:
+            raise ValueError("fraction_var must be in (0,1]")
+
+        if sv_tolerance is None:
+            pass
+        elif not isinstance(sv_tolerance, int):
+            raise TypeError("sv_tolerance must be an integer")
+        elif sv_tolerance <= 0:
+            raise ValueError("sv_tolerance must be greater than 0")
+
+        if n_components is None:
+            pass
+        elif not isinstance(n_components, int):
+            raise TypeError("n_components must be an integer")
+        elif n_components <= 0:
+            raise ValueError("n_components must be greater than 0")
+        elif n_components > min(size):
+            raise ValueError(
+                "n_components must be less than or equal to the \
+                    minimum input rank"
+            )
+
+    def fit(
```

Comment:
This function is pretty long. Is there a good way of splitting it up? Since all this code is numerical computing it's not too big a deal but good practice if you are able to.